File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/test/java/io/github/lishangbu/avalon/oauth2/authorizationserver/autoconfiguration/JWKSourceAutoConfigurationTest.java
package io.github.lishangbu.avalon.oauth2.authorizationserver.autoconfiguration;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.nimbusds.jose.jwk.JWKSet;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
import com.nimbusds.jose.jwk.source.JWKSource;
import com.nimbusds.jose.proc.SecurityContext;
import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.Base64;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.springframework.core.io.AbstractResource;
import org.springframework.core.io.DefaultResourceLoader;
import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.test.util.ReflectionTestUtils;

class JWKSourceAutoConfigurationTest {

    @TempDir Path tempDir;

    @Test
    void jwkSourceGeneratesKeysWhenNoLocationsProvided() {
        Oauth2Properties properties = new Oauth2Properties();
        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());

        configuration.afterPropertiesSet();
        JWKSource<SecurityContext> jwkSource = configuration.jwkSource();
        JWKSource<SecurityContext> secondCall = configuration.jwkSource();

        assertNotNull(jwkSource);
        assertNotNull(secondCall);
        JWKSet jwkSet = ((ImmutableJWKSet<SecurityContext>) jwkSource).getJWKSet();
        assertEquals(1, jwkSet.getKeys().size());
    }

    @Test
    void jwkSourceLoadsKeysFromConfiguredLocations() throws Exception {
        KeyPair keyPair = generateKeyPair();
        Path publicKeyFile = tempDir.resolve("public.pem");
        Path privateKeyFile = tempDir.resolve("private.pem");
        writePem(publicKeyFile, "PUBLIC KEY", keyPair.getPublic().getEncoded());
        writePem(privateKeyFile, "PRIVATE KEY", keyPair.getPrivate().getEncoded());

        Oauth2Properties properties = new Oauth2Properties();
        properties.setJwtPublicKeyLocation("file:" + publicKeyFile.toString());
        properties.setJwtPrivateKeyLocation("file:" + privateKeyFile.toString());

        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());
        configuration.afterPropertiesSet();

        JWKSet jwkSet = ((ImmutableJWKSet<SecurityContext>) configuration.jwkSource()).getJWKSet();
        RSAKey rsaKey = (RSAKey) jwkSet.getKeys().getFirst();

        assertEquals(
                ((RSAPublicKey) keyPair.getPublic()).getModulus(),
                rsaKey.toRSAPublicKey().getModulus());
        assertNotNull(rsaKey.toRSAPrivateKey());
    }

    @Test
    void afterPropertiesSetHandlesMissingResourcesAndInvalidContent() throws Exception {
        Oauth2Properties properties = new Oauth2Properties();
        properties.setJwtPublicKeyLocation("file:" + tempDir.resolve("missing.pub"));
        properties.setJwtPrivateKeyLocation("file:" + tempDir.resolve("missing.key"));

        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());
        configuration.afterPropertiesSet();

        Path invalidPublic = tempDir.resolve("invalid.pub");
        Path invalidPrivate = tempDir.resolve("invalid.key");
        Files.writeString(invalidPublic, "not-a-key", StandardCharsets.UTF_8);
        Files.writeString(invalidPrivate, "not-a-key", StandardCharsets.UTF_8);

        Oauth2Properties invalidProps = new Oauth2Properties();
        invalidProps.setJwtPublicKeyLocation("file:" + invalidPublic);
        invalidProps.setJwtPrivateKeyLocation("file:" + invalidPrivate);

        JWKSourceAutoConfiguration invalidConfig =
                new JWKSourceAutoConfiguration(invalidProps, new DefaultResourceLoader());
        invalidConfig.afterPropertiesSet();
        assertNotNull(invalidConfig.jwkSource());
    }

    @Test
    void afterPropertiesSetHandlesUnreadableResources() {
        Oauth2Properties properties = new Oauth2Properties();
        properties.setJwtPublicKeyLocation("unreadable:public");
        properties.setJwtPrivateKeyLocation("unreadable:private");

        ResourceLoader resourceLoader =
                new ResourceLoader() {
                    @Override
                    public Resource getResource(String location) {
                        return new AbstractResource() {
                            @Override
                            public String getDescription() {
                                return "unreadable";
                            }

                            @Override
                            public boolean exists() {
                                return true;
                            }

                            @Override
                            public boolean isReadable() {
                                return true;
                            }

                            @Override
                            public InputStream getInputStream() throws IOException {
                                throw new IOException("boom");
                            }
                        };
                    }

                    @Override
                    public ClassLoader getClassLoader() {
                        return getClass().getClassLoader();
                    }
                };

        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, resourceLoader);
        configuration.afterPropertiesSet();
    }

    @Test
    void afterPropertiesSetWarnsWhenOnlyOneKeyProvided() throws Exception {
        KeyPair keyPair = generateKeyPair();
        Path publicKeyFile = tempDir.resolve("public-only.pem");
        writePem(publicKeyFile, "PUBLIC KEY", keyPair.getPublic().getEncoded());

        Oauth2Properties properties = new Oauth2Properties();
        properties.setJwtPublicKeyLocation("file:" + publicKeyFile.toString());
        properties.setJwtPrivateKeyLocation(" ");

        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());
        configuration.afterPropertiesSet();
        assertNotNull(configuration.jwkSource());

        Oauth2Properties privateOnly = new Oauth2Properties();
        Path privateKeyFile = tempDir.resolve("private-only.pem");
        writePem(privateKeyFile, "PRIVATE KEY", keyPair.getPrivate().getEncoded());
        privateOnly.setJwtPublicKeyLocation(" ");
        privateOnly.setJwtPrivateKeyLocation("file:" + privateKeyFile.toString());

        JWKSourceAutoConfiguration privateConfig =
                new JWKSourceAutoConfiguration(privateOnly, new DefaultResourceLoader());
        privateConfig.afterPropertiesSet();
        assertNotNull(privateConfig.jwkSource());
    }

    @Test
    void generateRsaKeyThrowsWhenAlgorithmMissing() {
        Oauth2Properties properties = new Oauth2Properties();
        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());

        try (MockedStatic<KeyPairGenerator> mocked = Mockito.mockStatic(KeyPairGenerator.class)) {
            mocked.when(() -> KeyPairGenerator.getInstance("RSA"))
                    .thenThrow(new NoSuchAlgorithmException("RSA"));

            IllegalStateException exception =
                    assertThrows(
                            IllegalStateException.class,
                            () ->
                                    ReflectionTestUtils.invokeMethod(
                                            configuration, "generateRsaKey"));
            assertNotNull(exception.getMessage());
        }
    }

    @Test
    void usesFallbackKidWhenThumbprintComputationFails() throws Exception {
        KeyPair keyPair = generateKeyPair();
        Oauth2Properties properties = new Oauth2Properties();
        JWKSourceAutoConfiguration configuration =
                new JWKSourceAutoConfiguration(properties, new DefaultResourceLoader());
        ReflectionTestUtils.setField(
                configuration, "publicKey", (RSAPublicKey) keyPair.getPublic());
        ReflectionTestUtils.setField(
                configuration, "privateKey", (RSAPrivateKey) keyPair.getPrivate());

        try (MockedStatic<MessageDigest> mocked = Mockito.mockStatic(MessageDigest.class)) {
            mocked.when(() -> MessageDigest.getInstance("SHA-256"))
                    .thenThrow(new NoSuchAlgorithmException("SHA-256"));

            JWKSource<SecurityContext> jwkSource = configuration.jwkSource();
            assertNotNull(jwkSource);
        }
    }

    private static KeyPair generateKeyPair() throws Exception {
        KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
        generator.initialize(1024);
        return generator.generateKeyPair();
    }

    private static void writePem(Path file, String type, byte[] encoded) throws IOException {
        String content =
                "-----BEGIN "
                        + type
                        + "-----\n"
                        + Base64.getEncoder().encodeToString(encoded)
                        + "\n-----END "
                        + type
                        + "-----\n";
        Files.writeString(file, content, StandardCharsets.UTF_8);
    }
}

File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/test/kotlin/io/github/lishangbu/avalon/oauth2/authorizationserver/autoconfiguration/JWKSourceAutoConfigurationTest.kt
package io.github.lishangbu.avalon.oauth2.authorizationserver.autoconfiguration

import com.nimbusds.jose.jwk.RSAKey
import com.nimbusds.jose.jwk.source.ImmutableJWKSet
import com.nimbusds.jose.jwk.source.JWKSource
import com.nimbusds.jose.proc.SecurityContext
import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.mockito.MockedStatic
import org.mockito.Mockito
import org.springframework.core.io.AbstractResource
import org.springframework.core.io.DefaultResourceLoader
import org.springframework.core.io.Resource
import org.springframework.core.io.ResourceLoader
import org.springframework.test.util.ReflectionTestUtils
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.util.*

class JWKSourceAutoConfigurationTest {
    @TempDir
    lateinit var tempDir: Path

    @Test
    fun jwkSourceGeneratesKeysWhenNoLocationsProvided() {
        val properties = Oauth2Properties()
        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())

        configuration.afterPropertiesSet()
        val jwkSource: JWKSource<SecurityContext> = configuration.jwkSource()
        val secondCall: JWKSource<SecurityContext> = configuration.jwkSource()

        assertNotNull(jwkSource)
        assertNotNull(secondCall)
        val jwkSet = (jwkSource as ImmutableJWKSet<SecurityContext>).jwkSet
        assertEquals(1, jwkSet.keys.size)
    }

    @Test
    fun jwkSourceLoadsKeysFromConfiguredLocations() {
        val keyPair = generateKeyPair()
        val publicKeyFile = tempDir.resolve("public.pem")
        val privateKeyFile = tempDir.resolve("private.pem")
        writePem(publicKeyFile, "PUBLIC KEY", keyPair.public.encoded)
        writePem(privateKeyFile, "PRIVATE KEY", keyPair.private.encoded)

        val properties = Oauth2Properties()
        properties.jwtPublicKeyLocation = "file:$publicKeyFile"
        properties.jwtPrivateKeyLocation = "file:$privateKeyFile"

        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())
        configuration.afterPropertiesSet()

        val jwkSet = (configuration.jwkSource() as ImmutableJWKSet<SecurityContext>).jwkSet
        val rsaKey = jwkSet.keys.first() as RSAKey

        assertEquals((keyPair.public as RSAPublicKey).modulus, rsaKey.toRSAPublicKey().modulus)
        assertNotNull(rsaKey.toRSAPrivateKey())
    }

    @Test
    fun afterPropertiesSetHandlesMissingResourcesAndInvalidContent() {
        val properties = Oauth2Properties()
        properties.jwtPublicKeyLocation = "file:${tempDir.resolve("missing.pub")}"
        properties.jwtPrivateKeyLocation = "file:${tempDir.resolve("missing.key")}"

        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())
        configuration.afterPropertiesSet()

        val invalidPublic = tempDir.resolve("invalid.pub")
        val invalidPrivate = tempDir.resolve("invalid.key")
        Files.writeString(invalidPublic, "not-a-key", StandardCharsets.UTF_8)
        Files.writeString(invalidPrivate, "not-a-key", StandardCharsets.UTF_8)

        val invalidProps = Oauth2Properties()
        invalidProps.jwtPublicKeyLocation = "file:$invalidPublic"
        invalidProps.jwtPrivateKeyLocation = "file:$invalidPrivate"

        val invalidConfig = JWKSourceAutoConfiguration(invalidProps, DefaultResourceLoader())
        invalidConfig.afterPropertiesSet()
        assertNotNull(invalidConfig.jwkSource())
    }

    @Test
    fun afterPropertiesSetHandlesUnreadableResources() {
        val properties = Oauth2Properties()
        properties.jwtPublicKeyLocation = "unreadable:public"
        properties.jwtPrivateKeyLocation = "unreadable:private"

        val resourceLoader =
            object : ResourceLoader {
                override fun getResource(location: String): Resource =
                    object : AbstractResource() {
                        override fun getDescription(): String = "unreadable"

                        override fun exists(): Boolean = true

                        override fun isReadable(): Boolean = true

                        override fun getInputStream(): InputStream = throw IOException("boom")
                    }

                override fun getClassLoader(): ClassLoader = javaClass.classLoader
            }

        val configuration = JWKSourceAutoConfiguration(properties, resourceLoader)
        configuration.afterPropertiesSet()
    }

    @Test
    fun afterPropertiesSetWarnsWhenOnlyOneKeyProvided() {
        val keyPair = generateKeyPair()
        val publicKeyFile = tempDir.resolve("public-only.pem")
        writePem(publicKeyFile, "PUBLIC KEY", keyPair.public.encoded)

        val properties = Oauth2Properties()
        properties.jwtPublicKeyLocation = "file:$publicKeyFile"
        properties.jwtPrivateKeyLocation = " "

        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())
        configuration.afterPropertiesSet()
        assertNotNull(configuration.jwkSource())

        val privateOnly = Oauth2Properties()
        val privateKeyFile = tempDir.resolve("private-only.pem")
        writePem(privateKeyFile, "PRIVATE KEY", keyPair.private.encoded)
        privateOnly.jwtPublicKeyLocation = " "
        privateOnly.jwtPrivateKeyLocation = "file:$privateKeyFile"

        val privateConfig = JWKSourceAutoConfiguration(privateOnly, DefaultResourceLoader())
        privateConfig.afterPropertiesSet()
        assertNotNull(privateConfig.jwkSource())
    }

    @Test
    fun generateRsaKeyThrowsWhenAlgorithmMissing() {
        val properties = Oauth2Properties()
        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())

        val mocked: MockedStatic<KeyPairGenerator> =
            Mockito.mockStatic(KeyPairGenerator::class.java)
        mocked.use {
            it
                .`when`<KeyPairGenerator> { KeyPairGenerator.getInstance("RSA") }
                .thenThrow(NoSuchAlgorithmException("RSA"))

            val exception =
                assertThrows(IllegalStateException::class.java) {
                    ReflectionTestUtils.invokeMethod<Any>(configuration, "generateRsaKey")
                }
            assertNotNull(exception.message)
        }
    }

    @Test
    fun usesFallbackKidWhenThumbprintComputationFails() {
        val keyPair = generateKeyPair()
        val properties = Oauth2Properties()
        val configuration = JWKSourceAutoConfiguration(properties, DefaultResourceLoader())
        ReflectionTestUtils.setField(configuration, "publicKey", keyPair.public as RSAPublicKey)
        ReflectionTestUtils.setField(configuration, "privateKey", keyPair.private as RSAPrivateKey)

        val mocked: MockedStatic<MessageDigest> = Mockito.mockStatic(MessageDigest::class.java)
        mocked.use {
            it
                .`when`<MessageDigest> { MessageDigest.getInstance("SHA-256") }
                .thenThrow(NoSuchAlgorithmException("SHA-256"))

            val jwkSource: JWKSource<SecurityContext> = configuration.jwkSource()
            assertNotNull(jwkSource)
        }
    }

    companion object {
        private fun generateKeyPair(): KeyPair {
            val generator = KeyPairGenerator.getInstance("RSA")
            generator.initialize(1024)
            return generator.generateKeyPair()
        }

        private fun writePem(
            file: Path,
            type: String,
            encoded: ByteArray,
        ) {
            val content =
                "-----BEGIN $type-----\n" +
                    Base64.getEncoder().encodeToString(encoded) +
                    "\n-----END $type-----\n"
            Files.writeString(file, content, StandardCharsets.UTF_8)
        }
    }
}
